src/consensus/pool/finality_tracker.rs

```diff
@@ -0,0 +1,344 @@
+}
+
+impl FinalityTracker {
+    /// Adds the given `parent` for the given `block`.
+    ///
+    /// Handles possibly resulting implicit finalizations.
+    ///
+    /// Returns a [`FinalizationEvent`] that contains information about newly finalized slots.
+    pub fn add_parent(&mut self, block: BlockId, parent: BlockId) -> FinalizationEvent {
+        assert!(block.0 > parent.0);
+        if let Some(p) = self.parents.get(&block) {
+            assert!(p == &parent);
+            return FinalizationEvent::default();
+        }
+        self.parents.insert(block, parent);
+
+        let (slot, block_hash) = block;
+        let Some(status) = self.status.get(&slot) else {
+            return FinalizationEvent::default();
+        };
+        match status {
+            FinalizationStatus::FinalizedAndNotarized(hash)
+            | FinalizationStatus::ImplicitlyFinalized(hash) => {
+                let mut event = FinalizationEvent::default();
+                if &block_hash == hash {
+                    self.handle_implicitly_finalized(block.0, parent, &mut event);
+                }
+                event
+            }
+            FinalizationStatus::Notarized(_)
+            | FinalizationStatus::Finalized
+            | FinalizationStatus::ImplicitlySkipped => FinalizationEvent::default(),
+        }
+    }
+
+    /// Marks the given block as fast finalized.
+    ///
+    /// If the block was newly finalized, handles resulting implicit finalizations.
+    ///
+    /// Returns a [`FinalizationEvent`] that contains information about newly finalized slots.
+    pub fn mark_fast_finalized(&mut self, slot: Slot, block_hash: Hash) -> FinalizationEvent {
+        if let Some(status) = self.status.get(&slot) {
+            match status {
+                FinalizationStatus::FinalizedAndNotarized(_) => {
+                    return FinalizationEvent::default();
+                }
+                FinalizationStatus::Notarized(_)
+                | FinalizationStatus::Finalized
+                | FinalizationStatus::ImplicitlyFinalized(_) => {}
+                FinalizationStatus::ImplicitlySkipped => unreachable!("consensus safety violation"),
+            }
+        }
+
+        let mut event = FinalizationEvent::default();
+        self.status
+            .insert(slot, FinalizationStatus::FinalizedAndNotarized(block_hash));
+        self.handle_finalized((slot, block_hash), &mut event);
+        self.highest_finalized = slot.max(self.highest_finalized);
+        event
+    }
+
+    /// Marks the given block as notarized.
+    ///
+    /// Handles possibly resulting direct finalization of the block.
+    /// Further, also handles any possibly resulting implicit finalizations.
+    ///
+    /// Returns a [`FinalizationEvent`] that contains information about newly finalized slots.
+    pub fn mark_notarized(&mut self, slot: Slot, block_hash: Hash) -> FinalizationEvent {
+        if let Some(status) = self.status.get(&slot) {
```

Comment:
we are accessing the map a couple of different times in this function.  I wonder if the entries api will be more readable.